pkg/kvstore/etcd.go

```diff
@@ -24,51 +24,134 @@ import (
 	"github.com/cilium/cilium/common/types"
+	"github.com/cilium/cilium/pkg/controller"
 	"github.com/cilium/cilium/pkg/lock"
 	"github.com/cilium/cilium/pkg/logging/logfields"
-	"github.com/cilium/cilium/pkg/policy"
 
 	client "github.com/coreos/etcd/clientv3"
 	"github.com/coreos/etcd/clientv3/concurrency"
 	clientyaml "github.com/coreos/etcd/clientv3/yaml"
-	"github.com/coreos/etcd/mvcc/mvccpb"
 	"github.com/hashicorp/go-version"
 	"github.com/sirupsen/logrus"
 	ctx "golang.org/x/net/context"
 )
 
 const (
-	// eAddr is the string representing the key mapping to the value of the
-	// address for Etcd.
-	eAddr = "etcd.address"
-	// eCfg is the string representing the key mapping to the path of the
-	// configuration for Etcd.
-	eCfg = "etcd.config"
+	etcdName = "etcd"
+
+	addrOption = "etcd.address"
+	cfgOption  = "etcd.config"
 )
 
+type etcdModule struct {
+	opts   backendOptions
+	config *client.Config
+}
+
 var (
-	minEVersion, _ = version.NewConstraint(">= 3.1.0")
+	minRequiredVersion, _ = version.NewConstraint(">= 3.1.0")
+
+	// etcdDummyAddress can be overwritten from test invokers using ldflags
+	etcdDummyAddress = "http://127.0.0.1:4002"
+
+	etcdInstance = &etcdModule{
+		opts: backendOptions{
+			addrOption: &backendOption{
+				description: "Addresses of etcd cluster",
+			},
+			cfgOption: &backendOption{
+				description: "Path to etcd configuration file",
+			},
+		},
+	}
 )
 
-// EtcdOpts is the set of supported options for Etcd configuration.
-var EtcdOpts = map[string]bool{
-	eAddr: true,
-	eCfg:  true,
+func (e *etcdModule) getName() string {
+	return etcdName
+}
+
+func (e *etcdModule) setConfigDummy() {
+	e.config = &client.Config{}
+	e.config.Endpoints = []string{etcdDummyAddress}
+}
+
+func (e *etcdModule) setConfig(opts map[string]string) error {
+	return setOpts(opts, e.opts)
 }
 
-type EtcdClient struct {
-	cli         *client.Client
-	sessionMU   lock.RWMutex
+func (e *etcdModule) getConfig() map[string]string {
+	return getOpts(e.opts)
+}
+
+func (e *etcdModule) newClient() (BackendOperations, error) {
+	endpointsOpt, endpointsSet := e.opts[addrOption]
+	configPathOpt, configSet := e.opts[cfgOption]
+	configPath := ""
+
+	if e.config == nil {
+		if !endpointsSet && !configSet {
+			return nil, fmt.Errorf("invalid etcd configuration, %s or %s must be specified", cfgOption, addrOption)
+		}
+
+		e.config = &client.Config{}
+
+		if endpointsSet {
+			e.config.Endpoints = []string{endpointsOpt.value}
+		}
+
+		if configSet {
+			configPath = configPathOpt.value
+		}
+	}
+
+	return newEtcdClient(e.config, configPath)
+}
+
+func init() {
+	// register etcd module for use
+	registerBackend(etcdName, etcdInstance)
+}
+
+type etcdClient struct {
+	// protects all members of etcdClient from concurrent access
+	lock.RWMutex
+
+	client      *client.Client
 	session     *concurrency.Session
 	lockPathsMU lock.Mutex
 	lockPaths   map[string]*lock.Mutex
 }
 
-type EtcdLocker struct {
+type etcdMutex struct {
```

Comment:
I guess this was already like this but if it was `type etcdMutex concurrency.Mutex` I think it can avoid the pointer and the struct.

Reply:
Agreed, this is a non trivial change though because the Locker interface currently uses `Unlock()` and so does `concurrency.Mutex` but with a different function signature. I'm happy if someone cleans this up after.